OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,448 @@
+﻿/* 
+ * Copyright 2022, Optimizely
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using OptimizelySDK.Logger;
+using OptimizelySDK.Odp.Entity;
+using OptimizelySDK.Utils;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OptimizelySDK.Odp
+{
+    public class OdpEventManager : IOdpEventManager, IDisposable
+    {
+        /// <summary>
+        /// Enumeration of acceptable states of the Event Manager
+        /// </summary>
+        private enum ExecutionState
+        {
+            Stopped = 0,
+            Running = 1,
+            Processing = 2,
+        }
+
+        private const int MAX_RETRIES = 3;
+        private const int DEFAULT_BATCH_SIZE = 10;
+        private const int DEFAULT_FLUSH_INTERVAL_MSECS = 1000;
+        private const int DEFAULT_SERVER_QUEUE_SIZE = 10000;
+        public const string TYPE = "fullstack";
+
+        /// <summary>
+        /// Current state of the event processor
+        /// </summary>
+        private ExecutionState CurrentState
+        {
+            get
```

Comment:
I don't think this is the right place to `lock`
In the main logic, this should be `locked`,